src/util/KinBodyParser.cpp

```diff
@@ -0,0 +1,532 @@
+  }
+
+  bodyNodeInfo.properties = std::move(properties);
+  bodyNodeInfo.initTransform = initTransform;
+
+  return bodyNodeInfo;
+}
+
+//==============================================================================
+void readGeoms(
+    dart::dynamics::BodyNode* bodyNode,
+    tinyxml2::XMLElement* bodyEle,
+    const dart::common::Uri& baseUri,
+    const dart::common::ResourceRetrieverPtr& retriever)
+{
+  assert(bodyNode);
+
+  dart::utils::ElementEnumerator geomIterator(bodyEle, "geom");
+
+  if (!bodyEle->FirstChildElement("geom"))
+  {
+    dtwarn << "[KinBodyParser] KinBody document '"
+           << baseUri.toString() << "' does not contain any <Geom> element "
+           << "under <Body>. This body will have no shape.\n";
+  }
+
+  while (geomIterator.next())
+  {
+    auto geomEle = geomIterator.get();
+    readGeom(geomEle, bodyNode, baseUri, retriever);
+  }
+}
+
+//==============================================================================
+dart::dynamics::ShapePtr readMeshShape(
+    const std::string& fileName,
+    double uniScale,
+    const dart::common::Uri& baseUri,
+    const dart::common::ResourceRetrieverPtr& retriever)
+{
+  Eigen::Vector3d scale = Eigen::Vector3d::Constant(uniScale);
+
+  auto meshUri = dart::common::Uri::getRelativeUri(baseUri, fileName);
+  auto model = dart::dynamics::MeshShape::loadMesh(meshUri, retriever);
+
+  if (model)
+  {
+    return std::make_shared<dart::dynamics::MeshShape>(
+          scale, model, meshUri, retriever);
+  }
+  else
+  {
+    dterr << "[KinBodyParser] Fail to load model '" << fileName << "'.\n";
+    return nullptr;
+  }
+}
+
+//==============================================================================
+std::string resolveShapeName(
+    tinyxml2::XMLElement* geomEle,
+    dart::dynamics::BodyNode* bodyNode)
+{
+  std::string shapeNodeName;
+
+  if (dart::utils::hasAttribute(geomEle, "name"))
+  {
+    shapeNodeName = dart::utils::getAttributeString(geomEle, "name");
+  }
+  else
+  {
+    shapeNodeName = bodyNode->getName()
+        + " shape (" + std::to_string(bodyNode->getShapeNodes().size()) + ")";
+  }
+
+  return shapeNodeName;
+}
+
+//==============================================================================
+std::pair<std::string, double> resolveFileNameAndScale(
+    tinyxml2::XMLElement* geomEle,
+    const std::string& renderOrData)
+{
+  // <Render> or <Data> contains the relative mesh file path and optionally the
+  // mesh scale (e.g., <Render>my/mesh/file.stl<Render> or
+  // <Render>my/mesh/file.stl 0.25<Render>). If the scale is not provided then
+  // this function returns 1.0 by default.
+
+  std::pair<std::string, double> ret{"", 1.0};
+
+  const auto fileNameAndScale
+      = split(dart::utils::getValueString(geomEle, renderOrData), " ");
+  assert(!fileNameAndScale.empty());
+
+  ret.first = fileNameAndScale[0];
+
+  if (fileNameAndScale.size() > 1u)
+  {
+    ret.second = std::atof(fileNameAndScale[1].c_str());
```

Comment:
Check for a conversion error. I am not sure if this is possible with `atof`, so you may need to use a `stringstream` instead.